File: src/task/src/proxy/proxy.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "proxy.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <filesystem>

namespace olive
{

namespace
{

std::string shell_quote(const std::string &s)
{
	std::string out = "'";
	for (char c : s) {
		if (c == '\'') {
			out += "'\\''";
		} else {
			out += c;
		}
	}
	out += "'";
	return out;
}

/**
 * @brief Probes the source duration with the ffprobe next to ffmpeg
 *
 * Returns 0 when ffprobe is unavailable or the duration cannot be
 * determined, in which case the task simply reports no intermediate
 * progress.
 */
double probe_source_duration_seconds(const std::string &ffmpeg_path,
									 const std::string &source_filename)
{
	std::filesystem::path ffprobe =
		std::filesystem::path(ffmpeg_path).parent_path() / "ffprobe";
#if defined(_WIN32)
	ffprobe += ".exe";
#endif
	std::error_code ec;
	if (!std::filesystem::exists(ffprobe, ec)) {
		return 0.0;
	}

	std::string cmd = shell_quote(ffprobe.string()) +
					  " -v error -show_entries format=duration"
					  " -of default=noprint_wrappers=1:nokey=1 " +
					  shell_quote(source_filename);
	FILE *pipe = popen(cmd.c_str(), "r");
	if (!pipe) {
		return 0.0;
	}
	char buf[128] = { 0 };
	std::string output;
	while (fgets(buf, sizeof(buf), pipe)) {
		output += buf;
	}
	if (pclose(pipe) != 0) {
		return 0.0;
	}
	return strtod(output.c_str(), nullptr);
}

} // namespace

ProxyTask::ProxyTask(const OakCodecTaskRequest &request,
					 const oakcodec_proxy_params &params)
	: source_filename_(request.input_filename ? request.input_filename : "")
	, stream_index_(request.stream_index)
	, params_(params)
	, output_filename_(request.output_filename ? request.output_filename
											   : "")
{
	// Divider-based requests (width/height == 0) take the source fraction
	if (request.proxy_width > 0 && request.proxy_height > 0) {
		params_.width = request.proxy_width;
		params_.height = request.proxy_height;
		params_.divider = 1;
	}

	set_title("Generating Proxy " + source_filename_ + ":" +
			  std::to_string(stream_index_));
}

std::vector<std::string>
ProxyTask::build_arguments(const std::string &source_filename,
						   int stream_index,
						   const oakcodec_proxy_params &params,
						   const std::string &output_filename)
{
	std::string scale_filter;
	if (params.divider > 1) {
		// Fraction of the source resolution, rounded down to even dimensions
		// as required by yuv420p
		scale_filter = "scale=w=trunc(iw/" + std::to_string(params.divider) +
					   "/2)*2:h=trunc(ih/" + std::to_string(params.divider) +
					   "/2)*2";
	} else {
		scale_filter = "scale=w=" + std::to_string(params.width) +
					   ":h=" + std::to_string(params.height) +
					   ":force_original_aspect_ratio=decrease";
	}

	const std::string container_format =
		params.extension[0] ? params.extension : "mp4";

	std::vector<std::string> args;
	args.emplace_back("-y");
	// Report machine-readable progress on stdout for the task dialog
	args.emplace_back("-nostats");
	args.emplace_back("-progress");
	args.emplace_back("pipe:1");
	args.emplace_back("-i");
	args.emplace_back(source_filename);
	// Map the requested video stream first so it is stream 0 in the proxy
	args.emplace_back("-map");
	args.emplace_back("0:" + std::to_string(stream_index));

	if (params.include_audio) {
		// Keep the source audio (if any) so the proxy can also be used for
		// audio preview. Audio streams follow the video stream in source order.
		args.emplace_back("-map");
		args.emplace_back("0:a?");
		args.emplace_back("-c:a");
		args.emplace_back("aac");
		args.emplace_back("-b:a");
		args.emplace_back("128k");
	} else {
		args.emplace_back("-an");
	}

	args.emplace_back("-vf");
	args.emplace_back(scale_filter);
	args.emplace_back("-c:v");
	args.emplace_back("libx264");
	args.emplace_back("-preset");
	args.emplace_back(params.preset);
	args.emplace_back("-crf");
	args.emplace_back(std::to_string(params.crf));
	args.emplace_back("-pix_fmt");
	args.emplace_back("yuv420p");
	args.emplace_back("-movflags");
	args.emplace_back("+faststart");
	args.emplace_back("-f");
	args.emplace_back(container_format);
	args.emplace_back(output_filename);

	return args;
}

double ProxyTask::parse_progress(const std::string &line,
								 double duration_seconds)
{
	if (duration_seconds <= 0.0) {
		return -1.0;
	}

	int64_t out_time_us = -1;
	if (line.rfind("out_time_us=", 0) == 0) {
		out_time_us = strtoll(line.c_str() + 12, nullptr, 10);
	} else if (line.rfind("out_time_ms=", 0) == 0) {
		// Despite the name, ffmpeg reports this value in microseconds
		out_time_us = strtoll(line.c_str() + 12, nullptr, 10);
	}

	if (out_time_us < 0) {
		return -1.0;
	}

	double progress = out_time_us / 1000000.0 / duration_seconds;
	return progress < 0.0 ? 0.0 : (progress > 1.0 ? 1.0 : progress);
}

bool ProxyTask::run()
{
	char ffmpeg_buf[1024];
	if (oakcodec_proxy_find_ffmpeg(nullptr, ffmpeg_buf,
								   sizeof(ffmpeg_buf)) <= 0) {
		set_error("Failed to generate proxy: ffmpeg executable was not "
				  "found. Set the ffmpeg path in Preferences > Disk > Proxy "
				  "Settings.");
		fprintf(stderr, "ProxyTask: ffmpeg executable not found\n");
		return false;
	}

	std::filesystem::path output_dir =
		std::filesystem::path(output_filename_).parent_path();
	std::error_code ec;
	if (!output_dir.empty() && !std::filesystem::exists(output_dir, ec) &&
		!std::filesystem::create_directories(output_dir, ec)) {
		set_error("Failed to create proxy output directory");
		return false;
	}

	fprintf(stderr, "ProxyTask: starting ffmpeg proxy generation: %s -> %s\n",
			source_filename_.c_str(), output_filename_.c_str());

	std::filesystem::remove(output_filename_, ec);

	const std::string working_filename = output_filename_ + ".working.mp4";
	std::filesystem::remove(working_filename, ec);

	const std::vector<std::string> args = build_arguments(
		source_filename_, stream_index_, params_, working_filename);

	std::string cmd = shell_quote(ffmpeg_buf);
	for (const std::string &arg : args) {
		cmd += " ";
		cmd += shell_quote(arg);
	}
	cmd += " 2>&1";

	const double duration_seconds =
		probe_source_duration_seconds(ffmpeg_buf, source_filename_);

	FILE *process = popen(cmd.c_str(), "r");
	if (!process) {
		set_error("Failed to start ffmpeg for proxy generation");
		return false;
	}

	std::string progress_buffer;
	double last_progress = 0.0;
	char buf[256];

	while (true) {
		size_t n = fread(buf, 1, sizeof(buf), process);
		if (n > 0) {
			progress_buffer.append(buf, n);
			size_t newline;
			while ((newline = progress_buffer.find('\n')) !=
				   std::string::npos) {
				std::string line = progress_buffer.substr(0, newline);
				progress_buffer.erase(0, newline + 1);
				double progress = parse_progress(line, duration_seconds);
				if (progress >= 0.0 && progress - last_progress > 0.001) {
					last_progress = progress;
					emit_progress(progress);
				}
			}
		}

		if (n < sizeof(buf)) {
			if (feof(process)) {
				break;
			}
			if (ferror(process)) {
				break;
			}
		}

		if (is_cancelled()) {
			pclose(process);
			std::filesystem::remove(working_filename, ec);
			set_error("Proxy generation was cancelled");
			return false;
		}
	}

	int exit_status = pclose(process);
	if (exit_status != 0) {
		std::filesystem::remove(working_filename, ec);
		set_error("ffmpeg failed to generate proxy");
		fprintf(stderr, "ProxyTask: ffmpeg failed with exit status %d\n",
				exit_status);
		return false;
	}

	if (!std::filesystem::exists(working_filename, ec)) {
		set_error("ffmpeg finished but proxy file was not created");
		return false;
	}

	std::filesystem::rename(working_filename, output_filename_, ec);
	if (ec) {
		set_error("Failed to move proxy into place");
		return false;
	}

	fprintf(stderr, "ProxyTask: proxy generation succeeded: %s\n",
			output_filename_.c_str());
	emit_progress(1.0);
	return true;
}

}

File: crates/oaktask/src/proxy.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! `ProxyTask`, mirroring `src/task/src/proxy/proxy.h`.
//!
//! Transcodes a video stream to a proxy via an external `ffmpeg` process
//! (oakcodec kind `OAKCODEC_TASK_PROXY`). The two pure functions
//! [`ProxyTask::build_arguments`] and [`ProxyTask::parse_progress`] are
//! parity/golden candidates (see tests/parity_test.rs).
//!
//! CPP-PARITY: src/task/src/proxy/proxy.h

use std::io::{BufRead, BufReader};
use std::process::{Command, Stdio};

use crate::bridge;
use crate::error::{Error, Result};
use crate::handle::CHandle;
use crate::task::{Task, TaskBehavior};

/// The proxy parameters, mirroring `oakcodec_proxy_params` in
/// `include/codec/proxy.h` (kept as plain fields so no oakcodec handle is
/// needed to build arguments).
pub struct ProxyParams {
	/// Target width (0 = unspecified/divider-based).
	pub width: i32,
	/// Target height (0 = unspecified/divider-based).
	pub height: i32,
	/// Resolution divider.
	pub divider: i32,
	/// Proxy format version.
	pub version: i32,
	/// CRF for the proxy encode.
	pub crf: i32,
	/// Whether audio is included in the proxy.
	pub include_audio: bool,
	/// Output container extension.
	pub extension: String,
	/// ffmpeg preset name.
	pub preset: String,
}

/// A proxy transcode task: runs `ffmpeg` with arguments from
/// [`ProxyTask::build_arguments`], parsing progress lines via
/// [`ProxyTask::parse_progress`].
pub struct ProxyTask {
	/// The shared task base.
	pub base: Task,
	/// Source media filename.
	pub source_filename: String,
	/// Target stream index.
	pub stream_index: i32,
	/// Proxy parameters.
	pub params: ProxyParams,
	/// Output proxy filename.
	pub output_filename: String,
	/// Total media duration in seconds (for progress scaling).
	duration_seconds: f64,
}

impl ProxyTask {
	/// Build a proxy task from an oakcodec request and proxy params,
	/// mirroring the C++ constructor (divider-based requests take the source
	/// fraction).
	pub fn new(request: &bridge::codec::OakCodecTaskRequest, params: ProxyParams) -> ProxyTask {
		let source = unsafe { crate::ffi::taskhandle::cstr_to_string(request.input_filename) };
		let output = unsafe { crate::ffi::taskhandle::cstr_to_string(request.output_filename) };
		let mut params = params;
		if request.proxy_width > 0 && request.proxy_height > 0 {
			params.width = request.proxy_width;
			params.height = request.proxy_height;
			params.divider = 1;
		}
		let title = format!("Generating Proxy {}:{}", source, request.stream_index);
		ProxyTask {
			base: Task::new(&title, CHandle::null()),
			source_filename: source,
			stream_index: request.stream_index,
			params,
			output_filename: output,
			duration_seconds: 0.0,
		}
	}

	/// Build the `ffmpeg` command-line argument vector for the proxy
	/// transcode. Mirrors `build_arguments` in proxy.cpp; the exact ordering
	/// and flag spelling are parity-locked (golden test).
	///
	/// CPP-PARITY: src/task/src/proxy/proxy.cpp (build_arguments)
	pub fn build_arguments(
		source_filename: &str,
		stream_index: i32,
		params: &ProxyParams,
		output_filename: &str,
	) -> Vec<String> {
		let scale_filter = if params.divider > 1 {
			format!(
				"scale=w=trunc(iw/{}/2)*2:h=trunc(ih/{}/2)*2",
				params.divider, params.divider
			)
		} else {
			format!(
				"scale=w={}:h={}:force_original_aspect_ratio=decrease",
				params.width, params.height
			)
		};

		let container_format = if params.extension.is_empty() {
			"mp4".to_string()
		} else {
			params.extension.clone()
		};

		let mut args: Vec<String> = vec![
			"-y".to_string(),
			"-nostats".to_string(),
			"-progress".to_string(),
			"pipe:1".to_string(),
			"-i".to_string(),
			source_filename.to_string(),
			"-map".to_string(),
			format!("0:{stream_index}"),
		];

		if params.include_audio {
			args.extend([
				"-map".to_string(),
				"0:a?".to_string(),
				"-c:a".to_string(),
				"aac".to_string(),
				"-b:a".to_string(),
				"128k".to_string(),
			]);
		} else {
			args.push("-an".to_string());
		}

		args.extend([
			"-vf".to_string(),
			scale_filter,
			"-c:v".to_string(),
			"libx264".to_string(),
			"-preset".to_string(),
			params.preset.clone(),
			"-crf".to_string(),
			params.crf.to_string(),
			"-pix_fmt".to_string(),
			"yuv420p".to_string(),
			"-movflags".to_string(),
			"+faststart".to_string(),
			"-f".to_string(),
			container_format,
			output_filename.to_string(),
		]);

		args
	}

	/// Parse a single `ffmpeg -progress` output line into a progress value in
	/// 0.0..=1.0 given the total duration. Mirrors `parse_progress` in
	/// proxy.cpp (golden test).
	///
	/// CPP-PARITY: src/task/src/proxy/proxy.cpp (parse_progress)
	pub fn parse_progress(line: &str, duration_seconds: f64) -> Option<f64> {
		if duration_seconds <= 0.0 {
			return None;
		}

		let out_time_us: i64 = if let Some(rest) = line.strip_prefix("out_time_us=") {
			rest.trim().parse().unwrap_or(-1)
		} else if let Some(rest) = line.strip_prefix("out_time_ms=") {
			// Despite the name, ffmpeg reports this value in microseconds.
			rest.trim().parse().unwrap_or(-1)
		} else {
			-1
		};

		if out_time_us < 0 {
			return None;
		}

		let progress = out_time_us as f64 / 1_000_000.0 / duration_seconds;
		Some(if progress < 0.0 {
			0.0
		} else if progress > 1.0 {
			1.0
		} else {
			progress
		})
	}
}

impl TaskBehavior for ProxyTask {
	/// Spawn `ffmpeg` with the built arguments, feed `-progress` lines to
	/// [`ProxyTask::parse_progress`] and emit them as task progress.
	fn run(&mut self, task: &mut Task) -> Result<()> {
		let mut ffmpeg_buf = [0i8; 1024];
		let found = unsafe {
			bridge::codec::oakcodec_proxy_find_ffmpeg(
				std::ptr::null(),
				ffmpeg_buf.as_mut_ptr(),
				ffmpeg_buf.len() as i32,
			)
		};
		if found <= 0 {
			task.set_error(
				"Failed to generate proxy: ffmpeg executable was not found. Set the ffmpeg path in Preferences > Disk > Proxy Settings.",
			);
			return Err(Error::Failed("ffmpeg executable was not found".to_string()));
		}
		let ffmpeg_path = buf_to_string(&ffmpeg_buf);

		// Create the output directory if needed.
		if let Some(parent) = std::path::Path::new(&self.output_filename).parent() {
			if !parent.as_os_str().is_empty() && !parent.exists() {
				if std::fs::create_dir_all(parent).is_err() {
					task.set_error("Failed to create proxy output directory");
					return Err(Error::Failed(
						"Failed to create proxy output directory".to_string(),
					));
				}
			}
		}

		let _ = std::fs::remove_file(&self.output_filename);
		let working_filename = format!("{}.working.mp4", self.output_filename);
		let _ = std::fs::remove_file(&working_filename);

		let args = Self::build_arguments(
			&self.source_filename,
			self.stream_index,
			&self.params,
			&working_filename,
		);

		// Probe the source duration for progress scaling (0 when unknown).
		self.duration_seconds = probe_source_duration_seconds(&ffmpeg_path, &self.source_filename);

		let mut command = Command::new(&ffmpeg_path);
		command
			.args(&args)
			.stdout(Stdio::piped())
			.stderr(Stdio::piped());
		let mut child = match command.spawn() {
			Ok(child) => child,
			Err(_) => {
				task.set_error("Failed to start ffmpeg for proxy generation");
				return Err(Error::Failed("Failed to start ffmpeg".to_string()));
			}
		};

		// Drain stderr in a thread so a chatty ffmpeg cannot block us.
		let stderr = child.stderr.take();
		if let Some(stderr) = stderr {
			std::thread::spawn(move || {
				let _ = BufReader::new(stderr).lines().count();
			});
		}

		let stdout = child.stdout.take();
		let mut last_progress = 0.0_f64;
		if let Some(stdout) = stdout {
			let reader = BufReader::new(stdout);
			for line in reader.lines() {
				match line {
					Ok(line) => {
						if let Some(progress) = Self::parse_progress(&line, self.duration_seconds) {
							if progress >= 0.0 && progress - last_progress > 0.001 {
								last_progress = progress;
								task.emit_progress(progress);
							}
						}
					}
					Err(_) => break,
				}
				if task.is_cancelled() {
					let _ = child.kill();
					let _ = child.wait();
					let _ = std::fs::remove_file(&working_filename);
					task.set_error("Proxy generation was cancelled");
					return Err(Error::Cancelled);
				}
			}
		}

		let exit_status = child.wait();
		if !matches!(exit_status, Ok(status) if status.success()) {
			let _ = std::fs::remove_file(&working_filename);
			task.set_error("ffmpeg failed to generate proxy");
			return Err(Error::Failed("ffmpeg failed to generate proxy".to_string()));
		}

		if !std::path::Path::new(&working_filename).exists() {
			task.set_error("ffmpeg finished but proxy file was not created");
			return Err(Error::Failed(
				"ffmpeg finished but proxy file was not created".to_string(),
			));
		}

		if std::fs::rename(&working_filename, &self.output_filename).is_err() {
			task.set_error("Failed to move proxy into place");
			return Err(Error::Failed("Failed to move proxy into place".to_string()));
		}

		task.emit_progress(1.0);
		Ok(())
	}
}

/// Read a NUL-terminated C char buffer into a String.
fn buf_to_string(buf: &[i8]) -> String {
	let len = buf.iter().position(|&c| c == 0).unwrap_or(buf.len());
	unsafe {
		String::from_utf8_lossy(std::slice::from_raw_parts(buf.as_ptr() as *const u8, len))
			.into_owned()
	}
}

/// Probe the source duration via `ffprobe` next to ffmpeg; 0.0 when
/// unavailable (in which case no intermediate progress is reported).
fn probe_source_duration_seconds(ffmpeg_path: &str, source_filename: &str) -> f64 {
	let ffprobe = std::path::Path::new(ffmpeg_path).with_file_name("ffprobe");
	if !ffprobe.exists() {
		return 0.0;
	}
	let output = Command::new(&ffprobe)
		.args([
			"-v",
			"error",
			"-show_entries",
			"format=duration",
			"-of",
			"default=noprint_wrappers=1:nokey=1",
			source_filename,
		])
		.output();
	match output {
		Ok(out) if out.status.success() => {
			let text = String::from_utf8_lossy(&out.stdout);
			text.trim().parse().unwrap_or(0.0)
		}
		_ => 0.0,
	}
}
